fastchat/serve/monitor/rating_systems.py

```diff
@@ -0,0 +1,376 @@
+    loss = (
+        -((np.log(probs) * outcomes + np.log(1.0 - probs) * (1.0 - outcomes))).sum()
+        + reg_loss
+    )
+
+    error = outcomes - probs
+    grad = reg * params  # initialize the grad as the regularization grad
+    matchups_grads = -alpha * error
+    np.add.at(
+        grad[:n_competitors], matchups[:, [0, 1]], matchups_grads[:, None] * DIFF_MASK
+    )
+    grad[n_competitors:] -= np.dot(features.T, error)
+    return loss, grad
+
+
+# note on regularization:
+# default reg is to 0.5 since the LogisticRegression default is 1.0
+# in the original implementation, matchups were duplicated
+# that made the ratio of log loss to reg loss "twice as high"
+# in this non-duplicated version for parity we also reduce the reg by one half to match
+def fit_contextual_bt(
+    matchups,
+    features,
+    outcomes,
+    models,
+    idxs=None,
+    alpha=math.log(10.0),
+    reg=0.5,
+    tol=1e-6,
+):
+    n_features = features.shape[1]
+    n_models = len(models)
+    initial_params = np.zeros(n_models + n_features, dtype=np.float64)
+    half_reg = reg / 2.0
+
+    # sample idxs optionally allow for fitting on a bootstrap sample of the dataset
+    if idxs is not None:
+        matchups, features, outcomes = matchups[idxs], features[idxs], outcomes[idxs]
+
+    result = minimize(
+        fun=contextual_bt_loss_and_grad,
+        x0=initial_params,
+        args=(n_models, matchups, features, outcomes, alpha, reg, half_reg),
+        jac=True,
+        method="L-BFGS-B",
+        options={"disp": False, "maxiter": 100, "gtol": tol},
+    )
+    return result["x"]
+
+
+def compute_style_control(
+    df, alpha=math.log(10.0), reg=0.5, init_rating=1000.0, scale=400.0, tol=1e-6
+):
+    matchups, features, outcomes, models = preprocess_for_style(df)
+    ratings_params = fit_contextual_bt(
+        matchups,
+        features,
+        outcomes,
+        models=models,
+        alpha=alpha,
+        reg=reg,
+        tol=tol,
+    )
+    ratings = ratings_params[: len(models)]
+    params = ratings_params[len(models) :]
+    scaled_ratings = scale_and_offset(ratings, models, scale, init_rating)
+    scaled_ratings = pd.Series(scaled_ratings, index=models).sort_values(
+        ascending=False
+    )
+    return scaled_ratings, params
+
+
+def compute_bootstrap_style_control(
+    df,
+    num_round,
+    alpha=math.log(10.0),
+    reg=0.5,
+    init_rating=1000.0,
+    scale=400.0,
+    tol=1e-6,
+):
+    matchups, features, outcomes, models = preprocess_for_style(df)
+
+    contextual_bt_fn = partial(
+        fit_contextual_bt,
+        matchups,
+        features,
+        outcomes,
+        models,
+        alpha=alpha,
+        reg=reg,
+        tol=tol,
+    )
+
+    boot_idxs = np.random.randint(
+        low=0, high=matchups.shape[0], size=(num_round, matchups.shape[0])
+    )
+
+    # this one is still memory and cpu intensive so don't make too many processes
+    with mp.Pool(4) as pool:
```

Comment:
Let's

1. Add a progress bar here. I found the method below to work well.
`results = list(tqdm(pool.imap(contextual_bt_fn, boot_idxs), total=num_rounds))`

2. Allow the argparse to set the number of cpu. Something like
```
with mp.Pool(num_cpu) as pool:
    results = list(tqdm(pool.imap(contextual_bt_fn, boot_idxs), total=num_rounds))
```